mpc/client/src/main.rs

```diff
@@ -0,0 +1,628 @@
+            .send()
+            .await?
+            .error_for_status()?;
+        tx_status
+            .send(Status::UploadEnded(payload_id.clone()))
+            .expect("impossible");
+    }
+    pool.close().await?;
+    Ok(())
+}
+
+fn full<T: Into<Bytes>>(chunk: T) -> BoxBody {
+    Full::new(chunk.into())
+        .map_err(|never| match never {})
+        .boxed()
+}
+
+async fn handle(
+    lock: Arc<AtomicBool>,
+    tx_status: Sender<Status>,
+    latest_status: Arc<RwLock<Status>>,
+    req: hyper::Request<hyper::body::Incoming>,
+) -> Result<hyper::Response<BoxBody>, DynError> {
+    let response = |status, body| {
+        Ok(hyper::Response::builder()
+            .header(hyper::header::ACCESS_CONTROL_ALLOW_ORIGIN, "*")
+            .header(hyper::header::CONTENT_TYPE, "application/json")
+            .status(status)
+            .body(body)
+            .unwrap())
+    };
+    let file_response = |status, body| {
+        Ok(hyper::Response::builder()
+            .header(hyper::header::ACCESS_CONTROL_ALLOW_ORIGIN, "*")
+            .header(hyper::header::CONTENT_TYPE, "application/octet-stream")
+            .header(
+                hyper::header::CONTENT_DISPOSITION,
+                format!("attachment; filename={CONTRIB_SK_PATH}"),
+            )
+            .status(status)
+            .body(body)
+            .unwrap())
+    };
+    let response_empty = |status| response(status, BoxBody::default());
+    match (req.method(), req.uri().path()) {
+        (&Method::POST, SK_ENDPOINT) => {
+            let whole_body = req.collect().await?.aggregate();
+            let email = serde_json::from_reader(whole_body.reader())?;
+            let guard = latest_status.write().await;
+            let result = {
+                if let Err(_) = tokio::fs::metadata(CONTRIB_SK_PATH).await {
+                    let secret_key = generate_pgp_key(email);
+                    let secret_key_serialized = secret_key
+                        .to_armored_bytes(ArmorOptions::default())
+                        .expect("impossible");
+                    tokio::fs::write(CONTRIB_SK_PATH, &secret_key_serialized).await?;
+                    response_empty(hyper::StatusCode::CREATED)
+                } else {
+                    response_empty(hyper::StatusCode::OK)
+                }
+            };
+            drop(guard);
+            result
+        }
+        (&Method::GET, SK_ENDPOINT) => {
+            if let Err(_) = tokio::fs::metadata(CONTRIB_SK_PATH).await {
+                response_empty(hyper::StatusCode::NOT_FOUND)
+            } else {
+                let content = tokio::fs::read(CONTRIB_SK_PATH).await?;
+                file_response(hyper::StatusCode::OK, full(content))
+            }
+        }
+        (&Method::POST, CONTRIBUTE_ENDPOINT)
+            if lock
+                .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+                .is_ok() =>
```

Comment:
writing to an atomic bool in an if guard is interesting, i would recommend pulling this out into the arm expression for clarity/ least surprise